File: Go/sorted_array_to_bst.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func toBST(nums []int, start, end int) *TreeNode {
    if start > end {
        return nil
    }
    mid := start + (end - start) / 2
    root := &TreeNode{Val: nums[mid]}
    
    root.Left = toBST(nums, start, mid - 1)
    root.Right = toBST(nums, mid + 1, end)

    return root
}

func sortedArrayToBST(nums []int) *TreeNode {
    if nums == nil || len(nums) == 0 {
        return nil
    }

    return toBST(nums, 0, len(nums) - 1)
}

File: Java/SortedArrayToBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    public TreeNode sortedArrayToBST(int[] nums, int start, int end) {
        if(start > end) return null;
        int mid = start + (end - start) / 2;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = sortedArrayToBST(nums, start, mid - 1);
        root.right = sortedArrayToBST(nums, mid + 1, end);

        return root;
    }

    public TreeNode sortedArrayToBST(int[] nums) {
        if(nums == null || nums.length == 0) return null;

        return sortedArrayToBST(nums, 0, nums.length - 1);
    }
}